File: rerun_cpp/src/rerun/archetypes/ellipsoids_ext.cpp
#include "ellipsoids.hpp"

#include "../collection_adapter_builtins.hpp"

// #define EDIT_EXTENSION

namespace rerun {
    namespace archetypes {

#ifdef EDIT_EXTENSION
        // <CODEGEN_COPY_TO_HEADER>

        /// Creates new `Ellipsoids` with `half_sizes` centered around the local origin.
        static Ellipsoids from_half_sizes(Collection<components::HalfSize3D> half_sizes) {
            Ellipsoids ellipsoids;
            ellipsoids.half_sizes = std::move(half_sizes);
            return ellipsoids;
        }

        /// Creates new `Ellipsoids` with `half_sizes` created from radii.
        //
        // TODO(andreas): This should not take an std::vector.
        static Ellipsoids from_radii(const std::vector<float>& sizes);

        /// Creates new `Ellipsoids` with `centers` and `half_sizes`.
        static Ellipsoids from_centers_and_half_sizes(
            Collection<components::Position3D> centers,
            Collection<components::HalfSize3D> half_sizes
        ) {
            Ellipsoids ellipsoids;
            ellipsoids.half_sizes = std::move(half_sizes);
            ellipsoids.centers = std::move(centers);
            return ellipsoids;
        }

        /// Creates new `Ellipsoids` with `half_sizes` and `centers` created from centers and radii.
        //
        // TODO(andreas): This should not take an std::vector.
        static Ellipsoids from_centers_and_radii(
            const std::vector<datatypes::Vec3D>& centers, const std::vector<float>& radii
        );

        // </CODEGEN_COPY_TO_HEADER>
#endif
        Ellipsoids Ellipsoids::from_radii(const std::vector<float>& radii) {
            std::vector<components::HalfSize3D> half_sizes;
            half_sizes.reserve(radii.size());
            for (const auto& radius : radii) {
                half_sizes.emplace_back(radius, radius, radius);
            }

            // Move the vector into a component batch.
            return Ellipsoids::from_half_sizes(std::move(half_sizes));
        }

        Ellipsoids Ellipsoids::from_centers_and_radii(
            const std::vector<datatypes::Vec3D>& centers, const std::vector<float>& radii
        ) {
            auto num_components = std::min(centers.size(), radii.size());

            std::vector<components::HalfSize3D> half_sizes;
            half_sizes.reserve(num_components);

            for (size_t i = 0; i < num_components; ++i) {
                float radius = radii[i];
                half_sizes.emplace_back(radius, radius, radius);
            }

            // We only transformed the radii; the centers are good as-is.
            Ellipsoids ellipsoids;
            ellipsoids.half_sizes = std::move(half_sizes);
            ellipsoids.centers = std::move(centers);
            return ellipsoids;
        }
    } // namespace archetypes
} // namespace rerun

File: crates/store/re_types/src/archetypes/ellipsoids_ext.rs
use crate::components::{HalfSize3D, Position3D};

use super::Ellipsoids;

impl Ellipsoids {
    /// Creates a new [`Ellipsoids`] for spheres with the given radii.
    // Note: This is not a `Radius` component because the `Radius` component is for
    // the on-screen sizes of lines and points.
    #[inline]
    #[doc(alias = "sphere")]
    pub fn from_radii(radii: impl IntoIterator<Item = f32>) -> Self {
        Self::new(radii.into_iter().map(HalfSize3D::splat))
    }

    /// Creates a new [`Ellipsoids`] for spheres with the given [`Self::centers`], and
    /// [`Self::half_sizes`] all equal to the given radii.
    // Note: This is not a `Radius` component because the `Radius` component is for
    // the on-screen sizes of lines and points.
    #[doc(alias = "sphere")]
    #[inline]
    pub fn from_centers_and_radii(
        centers: impl IntoIterator<Item = impl Into<Position3D>>,
        radii: impl IntoIterator<Item = f32>,
    ) -> Self {
        Self::new(radii.into_iter().map(HalfSize3D::splat)).with_centers(centers)
    }

    /// Creates a new [`Ellipsoids`] with [`Self::half_sizes`].
    #[inline]
    pub fn from_half_sizes(half_sizes: impl IntoIterator<Item = impl Into<HalfSize3D>>) -> Self {
        Self::new(half_sizes)
    }

    /// Creates a new [`Ellipsoids`] with [`Self::centers`] and [`Self::half_sizes`].
    #[inline]
    pub fn from_centers_and_half_sizes(
        centers: impl IntoIterator<Item = impl Into<Position3D>>,
        half_sizes: impl IntoIterator<Item = impl Into<HalfSize3D>>,
    ) -> Self {
        Self::new(half_sizes).with_centers(centers)
    }
}
